[flang][OpenMP] Convert unique clauses in ClauseProcessor

kparzysz opened the pull request:
Temporarily rename old clause list to `clauses2`, old clause iterator to `ClauseIterator2`.
Change `findUniqueClause` to iterate over `omp::Clause` objects, modify all handlers to operate on 'omp::clause::xyz` equivalents.

[Clause representation 2/6]

skatrak commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -2639,54 +2641,50 @@ bool ClauseProcessor::processFinal(Fortran::lower::StatementContext &stmtCtx,
-    if (orderedClause->v.has_value()) {
-      const auto *expr = Fortran::semantics::GetExpr(orderedClause->v);
-      orderedClauseValue = *Fortran::evaluate::ToInt64(*expr);
+    if (clause->v.has_value()) {
```
Nit: Braces not necessary.

kparzysz replied:
Done

skatrak reviewed (commented):
Thanks, LGTM.

kparzysz committed:
[flang][OpenMP] Implement flexible OpenMP clause representation

Introduce a set of generic classes (templates) that represent OpenMP
clauses in a language-agnostic manner. OpenMP clauses can contain
expressions and data objects and the exact representation of each
depends on the source language of the compiled program. To deal with
this, the templates depend on two type parameters:
- IdType: type that represent object's identity (in a way that
  satisfied OpenMP requirements), and
- ExprType: type that can represent numeric values, as well as
  data references (e.g. x.y[1].z[2]).

In addition to that, implement code instantiating these templates
from flang's AST.

This patch only introduces the new classes, they are not yet used
anywhere.

kparzysz committed:
[flang][OpenMP] Convert unique clauses in ClauseProcessor

Temporarily rename old clause list to `clauses2`, old clause iterator
to `ClauseIterator2`.
Change `findUniqueClause` to iterate over `omp::Clause` objects,
modify all handlers to operate on 'omp::clause::xyz` equivalents.

kiranchandramohan reviewed (approved):
Mostly mechanical changes. LGTM. Please wait for @skatrak

skatrak reviewed (approved):
LGTM.

tblah reviewed (approved):
LGTM

kparzysz committed:
Merge branch 'main' into users/kparzysz/spr/b02-unique
